### DISCUSSION TITLE
[BUG] Cannot get file versions for files greater than 2GB

### DISCUSSION THREAD

**Discussion by sosandu** [pnp/powershell#2696] [Q&A]
Cannot retrieve file versions for files greater than 2GB (Int32.MaxValue)
Get-PnPFileVersion -Url "/sites/modernteam/Shared Documents/3GB.docx"
Get-PnPFileVersion : Operation is not valid due to the current state of the object.
The property Size is obsolete and has a max of int32.maxvalue.
We need to use Length property instead so we should not use IncludeWithDefaultProperties
Same issue for Remove-PnPFileVersion

**Comment by alekpavlov** [comment]
How to use **Length** property with **Remove-PnPFileVersion**?

I'm trying to delete a version of a file which 9 GB in size and I'm getting the same error: **Operation is not valid due to the current state of the object.**

**Comment by frankchen76** [comment]
Deep dive in the [GetFileVersion.cs](https://github.com/pnp/powershell/blob/68da9f5cde4cabe5f870e0e8475478b07f49856d/src/Commands/Files/GetFileVersion.cs), [RemoveFileVersion.cs](https://github.com/pnp/powershell/blob/68da9f5cde4cabe5f870e0e8475478b07f49856d/src/Commands/Files/RemoveFileVersion.cs) and found out that the exception was raised at the following code. 
```
            ClientContext.Load(file, f => f.Exists, f => f.Versions.IncludeWithDefaultProperties(i => i.CreatedBy));
            ClientContext.ExecuteQueryRetry();
```
I am not sure if this is limit when file is over 2GB. Since both cmdlets used the same logic, Remove-PnPFile cmdlet will have this issue too.  @alekpavlov, I run the following PS which bypass the load process first. It works fine to delete all versions. 
```
# get file version using pnp PowerShell
$relativePath = "/sites/FrankCommunication1/Shared Documents/112233.docx"
$context = Get-PnPContext
$file = $context.Web.GetFileByServerRelativeUrl($relativePath)

# load file object without versions property
$context.Load($file)
$context.ExecuteQuery()

#delete all versions
$file.Versions.DeleteAll()
$context.ExecuteQuery()
```
I guess Remove-PnPFileVersion probably didn't need to load all versions if parameter indicated -All.

**Reply by alekpavlov** [reply]
Thank you @frankchen76 I hope this will be fixed in Remove-PnPFileVersion in the near future.

While testing I found out that file versions bigger than 2 GB could be deleted with the following workaround:
-----------------------------------------------------------------
```
$site1 = "https://contoso.sharepoint.com/"
$filepath1 = "/sites/Contoso/Test/Demo/Purpose/DemoVideo.mov "
Connect-PnPOnline -Url $site1 -Interactive
$file1 = Get-PnPFile -Url $filepath1 -AsFileObject 
$file1.Versions.DeleteAll() 
or 
$file1.Versions.DeleteByLabel(1.0) - if you want to delete a single version 
Invoke-PnPQuery
```
---------------------------------------------------------------

**Reply by jbcmd4** [reply]
Any chance you have figured out a way to delete all but a certain number of versions? I was trying to use get-pnpfile and get-pnpproperty but I guess I'm running into the same issues. However I don't want to delete every version just a certain amount. Thank you

**Reply by Hayholten** [reply]
Hello @jbcmd4 ,

I myself have had a lot of problems with the number of versions and file sizes.

After many nights of trying and searching, I finally managed to get the result I wanted. Don't hesitate to refer to this post where I detail the solution I found:

[https://stackoverflow.com/a/78028640](https://stackoverflow.com/a/78028640)

**Reply by Rootax** [reply]
> Hello @jbcmd4 ,
> 
> I myself have had a lot of problems with the number of versions and file sizes.
> 
> After many nights of trying and searching, I finally managed to get the result I wanted. Don't hesitate to refer to this post where I detail the solution I found:
> 
> https://stackoverflow.com/a/78028640

Nice work !

I try to work with this, but to only delete versions older than 720 days, without success :/  Any help would be very appreciated.

**Comment by jbcmd4** [comment]
I seen the solution but it's a all or nothing. I was hoping to find something where I could keep a certain amount of versions. Lets say there is 500 versions I want to keep 150.  In my case it appears only be a issue with files over the 2GB so lucky most of mine are below that.  Thank your insights.

Jon Benson
Collaboration Architect
Global Technology Services (GTS)
Schneider Digital
D  615-814-4107
M  615-545-5485
MS Teams IM  ***@***.******@***.***>
E  ***@***.******@***.***>
6700 Tower Cir #700
37172
Franklin,Tn
United States

***@***.***<https://youtu.be/dux6kG13QWM>

[Blog]<https://www.se.com/en/work/support/myschneider-app/>
Download mySchneider app<https://www.se.com/en/work/support/myschneider-app/>
24/7 support. Mobile catalog. Access to expert help.
*Please consider the environment before printing this e-mail
[Blog]<https://www.facebook.com/SchneiderElectric> [Blog] <Schneider-Electric>  [Blog] <https://blog.se.com/>  [Blog] <https://www.twitter.com/SchneiderElec>  [Blog] <https://www.youtube.com/schneidercorporate>  [Blog] <https://instagram.com/schneiderelectric>




General

**Comment by adambrubaker** [comment]
Attempting to give this thread a bump so that someone might look for a solution.  We are trying something slightly different.  Instead of deleting versions we are looking to get each version and the size to build a total size for each file.  Then we can look at the files taking up large amounts of space and make intelligent decisions over which ones to delete. 

Same issue.  When I run Get-PnPFileVersion on a file > 2GB I get "Operation is not valid due to the current state of the object."

**Comment by sabey2000** [comment]
Yup, ok I got the same problem this morning on one of the file on my sharepoint site.  I have an Azure logic app that generate report to the site owner that gives the REAL size used by a file with a kind of version factor, so they that a file a 10Gb with 200 version is not 10GB but something close to 2TB :P so they have to cleanup some versions and empty the 1st recycle bin to avoid being lock with the site 100% full.

Pretty sure it's another story of INT instead of INT64 since the API call work fine.  I will take a look in the PnP code in debug and see if I can give a solution
